network/crl_test.go

```diff
@@ -0,0 +1,468 @@
+		revokedCert:           TestRevokedCertFilename,
+		crl:                   TestCRLFilename,
+		ocspCert:              TestOCSPCertFilename,
+		ocspKey:               TestOCSPKeyFilename,
+		validRawCerts:         validRawCerts,
+		validVerifiedChains:   validVerifiedChains,
+		invalidRawCerts:       invalidRawCerts,
+		invalidVerifiedChains: invalidVerifiedChains,
+	}
+}
+
+// func getTestCertGroup3(t *testing.T) TestCertGroup {
+// 	validRawCerts, validVerifiedChains := getValidTestChain3(t)
+// 	invalidRawCerts, invalidVerifiedChains := getInvalidTestChain3(t)
+//
+// 	return TestCertGroup{
+// 		prefix:                TestCertPrefix3,
+// 		ca:                    TestCACertFilename3,
+// 		ica:                   TestICACertFilename3,
+// 		validCert:             TestCertFilename3,
+// 		revokedCert:           TestRevokedCertFilename3,
+// 		crl:                   TestCRLFilename3,
+// 		ocspCert:              TestOCSPCertFilename3,
+// 		ocspKey:               TestOCSPKeyFilename3,
+// 		validRawCerts:         validRawCerts,
+// 		validVerifiedChains:   validVerifiedChains,
+// 		invalidRawCerts:       invalidRawCerts,
+// 		invalidVerifiedChains: invalidVerifiedChains,
+// 	}
+// }
+
+func getTestCert(t *testing.T, filename string) ([]byte, *x509.Certificate) {
+	rawCert, err := ioutil.ReadFile(filename)
+	if err != nil {
+		t.Fatalf("Cannot read certificate: %v\n", err)
+	}
+
+	cert, err := pemToX509Cert(rawCert)
+	if err != nil {
+		t.Fatalf("Failed to parse certificate: %v\n", err)
+	}
+
+	return rawCert, cert
+}
+
+func getTestCRL(t *testing.T, filename string) ([]byte, *pkix.CertificateList) {
+	rawCRL, err := ioutil.ReadFile(filename)
+	if err != nil {
+		t.Fatalf("Cannot read CRL: %v\n", err)
+	}
+
+	crl, err := x509.ParseCRL(rawCRL)
+	if err != nil {
+		t.Fatalf("Failed to parse CRL: %v\n", err)
+	}
+
+	return rawCRL, crl
+}
+
+// Convert PEM-encoded certificate into DER and parse it into x509.Certificate
+func pemToX509Cert(data []byte) (*x509.Certificate, error) {
+	certDERBlock, _ := pem.Decode(data)
+	if certDERBlock == nil {
+		return nil, errors.New("Failed to decode PEM")
+	}
+
+	if certDERBlock.Type != "CERTIFICATE" {
+		return nil, errors.New("Decoded DER block was not a certificate")
+	}
+
+	cert, err := x509.ParseCertificate(certDERBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+// Starts HTTP server on some random port, uses `mux` to handle requests,
+// returns created server and IP:port of listening socket
+func getTestHTTPServer(t *testing.T, mux *http.ServeMux) (*http.Server, string) {
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatalf("Cannot create TCP socket for HTTP server: %v", err)
+	}
+
+	addr := listener.Addr().String()
+
+	httpServer := &http.Server{
+		Handler: mux,
+	}
+	go func() {
+		httpServer.Serve(listener)
+	}()
+
+	// Give the server time to start, otherwise this test will fail sometimes
+	time.Sleep(time.Millisecond * 100)
```

Comment:
Why is this necessary, given that the listener socket is already created? Any connections should be queued until the server goes up.

Reply:
When I was writing these tests, there were some issues (like client was unable to connect so I suspected server didn't have time to start), but now I don't observe this issue, so removed these lines. If the problem will appear again, we'll discuss it.